### PR TITLE
MODE-2551 Adds EPUB sequencer

### PR DESCRIPTION
**Pull request by RobSis**
Initial implementation of EPUB sequencer.

### CODE REVIEW

File: sequencers/modeshape-sequencer-epub/src/main/java/org/modeshape/sequencer/epub/EpubMetadata.java [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+                            date = node.getTextContent();
+                        }
+                    }
+                }
+            }
+            zip.close();
+
+            return true;
+        } catch (Exception e) {
+            // log
```
Comment (hchiorean) [inline sequencers/modeshape-sequencer-epub/src/main/java/org/modeshape/sequencer/epub/EpubMetadata.java]: the comment is correct, this should be logged or left to bubble up and logged in the sequencer

File: sequencers/modeshape-sequencer-epub/src/main/java/org/modeshape/sequencer/epub/EpubMetadata.java [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+
+    private InputStream in;
+    private DataInput din;
+
+    /*
+     * Check that given file is supported by this sequencer.
+     */
+    public boolean check() {
+        try {
+            ZipInputStream zip = new ZipInputStream(in);
```
Comment (hchiorean) [inline sequencers/modeshape-sequencer-epub/src/main/java/org/modeshape/sequencer/epub/EpubMetadata.java]: shouldn't this use try-with-resources ?

File: sequencers/modeshape-sequencer-epub/src/main/java/org/modeshape/sequencer/epub/EpubMetadata.java [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+    }
+
+
+    /**
+     * Set the input stream to the argument stream (or file). Note that {@link java.io.RandomAccessFile} implements
+     * {@link java.io.DataInput}.
+     *
+     * @param dataInput the input stream to read from
+     */
+    public void setInput( DataInput dataInput ) {
```
Comment (hchiorean) [inline sequencers/modeshape-sequencer-epub/src/main/java/org/modeshape/sequencer/epub/EpubMetadata.java]: is this used anywhere ? it seems only the `setInputStream(InputStream)` is used

**Comment by hchiorean** [comment]
The `EpubMetadata` seems to only be used together with an `InputStream`. As such,
a) its constructor should be changed and require this input stream (and hence no more setter)
b) the `DataInput` member should be removed if not used

Also, see [this comment](https://github.com/ModeShape/modeshape/pull/1499#commitcomment-15260782) on things that are still required for this to be included in the ModeShape distribution.

File: sequencers/modeshape-sequencer-epub/src/main/java/org/modeshape/sequencer/epub/EpubMetadata.java [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+
+    /*
+     * Check that given file is supported by this sequencer.
+     */
+    public boolean check() {
+        try {
+            ZipInputStream zip = new ZipInputStream(in);
+            ZipEntry entry = null;
+            while ((entry = zip.getNextEntry()) != null) {
+                if (entry.getName().endsWith("/content.opf")) {
```
Comment (maschmid) [inline sequencers/modeshape-sequencer-epub/src/main/java/org/modeshape/sequencer/epub/EpubMetadata.java]: I don't believe the EPUB spec requires the rootfile to be named "content.opf"

The name of the rootfile (or rootfiles) is definied in META-INF/container.xml  (see http://www.idpf.org/epub/301/spec/epub-ocf.html#sec-container-metainf-container.xml )

**Comment by hchiorean** [comment]
any status update on this ? ModeShape 4.6.Final will be released next week and most likely will be the last 4.x release which will include new features.

Everything else will get moved to ModeShape 5.

**Comment by RobSis** [comment]
Thanks for the input. 
I'm fine with moving it to v5, since I won't have too much time to work on it in next few weeks.

**Comment by hchiorean** [comment]
@RobSis: are there any plans of updating this in the near future ? We're evaluating the state of our open enhancement requests so we need to know whether to consider or close this

**Comment by RobSis** [comment]
Hello. Yes! 
Didn't have a lot of time, but I'm working on it and it should be ready in next week. 

Thank you for notice!

**Pull request closed without merging**